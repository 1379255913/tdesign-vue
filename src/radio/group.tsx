import Vue, { VNode, CreateElement } from 'vue';
import isString from 'lodash/isString';
import isNumber from 'lodash/isNumber';
import props from './radio-group-props';
import { RadioOptionObj, RadioOption, RadioValue } from './type';
import { prefix } from '../config';
import Radio, { radioBtnName } from './radio';
import { TNodeReturnValue } from '../common';
import CLASSNAMES from '../utils/classnames';

const name = `${prefix}-radio-group`;

export default Vue.extend({
  name: 'TRadioGroup',
  props: { ...props },

  components: {
    Radio,
  },

  provide(): Record<string, any> {
    return {
      radioGroup: this,
    };
  },

  data() {
    return {
      barStyle: {},
      observer: null,
    };
  },

  render(h: CreateElement): VNode {
    const { $scopedSlots } = this;
    let children: TNodeReturnValue = $scopedSlots.default && $scopedSlots.default(null);

    if (this.options && this.options.length) {
      children = this.options.map((option: RadioOption, index) => {
        let opt = option as RadioOptionObj;
        if (isNumber(option) || isString(option)) {
          opt = { value: option, label: option.toString() };
        }
        return (
          <Radio
            key={`radio-group-options-${opt.value}-${index}`}
            name={this.name}
            checked={this.value === opt.value}
            disabled={'disabled' in opt ? opt.disabled : this.disabled}
            value={opt.value}
          >
            {typeof opt.label === 'function' ? opt.label(h) : opt.label}
          </Radio>
        );
      });
    }

    // TODO 移除 buttonStyle api, 改用 variant
    if (this.buttonStyle === 'solid') {
      console.error('TDesign Radio Warn: buttonStyle will be deprecated, please use `variant` instead.');
    }
    const groupClass = [`${name}`, `${name}-${this.size}`, {
      [`${name}-outline`]: this.variant === 'outline',
      [`${name}-filled`]: this.buttonStyle === 'solid' || this.variant.includes('filled'),
      [`${name}-primary-filled`]: this.variant === 'primary-filled',
    }];
    if (this.buttonStyle === 'solid' || this.variant.includes('filled')) {
      children && children.push(<div style={this.barStyle} class={`${name}-filled-bg-block`}></div>);
    }

    return <div class={groupClass}>{children}</div>;
  },

  watch: {
    value() {
      this.$nextTick(() => this.calcBarStyle());
    },
  },

  mounted() {
    this.calcBarStyle();

    const observer = new MutationObserver(this.calcBarStyle);
    observer.observe(this.$el, { childList: true, attributes: true, subtree: true });
    this.observer = observer;
  },

  beforeDestroy() {
    this.observer.disconnect();
  },

  methods: {
    handleRadioChange(value: RadioValue, context: { e: Event }) {
      this.$emit('change', value, context);
      typeof this.onChange === 'function' && this.onChange(value, context);
    },
    calcBarStyle() {
      if (this.buttonStyle !== 'solid' && this.variant === 'outline') return;

      const checkedRadio: HTMLElement = this.$el.querySelector(`.${radioBtnName}.${CLASSNAMES.STATUS.checked}`);
      if (!checkedRadio) return;
      const { offsetWidth, offsetLeft } = checkedRadio;
      this.barStyle = { width: `${offsetWidth}px`, left: `${offsetLeft}px` };
    },
  },
});
