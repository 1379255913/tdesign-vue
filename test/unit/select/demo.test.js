import { mount } from '@vue/test-utils';
import base from '@/examples/select/demos/base.vue';
import collapsed from '@/examples/select/demos/collapsed.vue';
import creatable from '@/examples/select/demos/creatable.vue';
import customOptions from '@/examples/select/demos/custom-options.vue';
import customSelected from '@/examples/select/demos/custom-selected.vue';
import disabled from '@/examples/select/demos/disabled.vue';
import filterable from '@/examples/select/demos/filterable.vue';
import group from '@/examples/select/demos/group.vue';
import labelInValue from '@/examples/select/demos/label-in-value.vue';
import max from '@/examples/select/demos/max.vue';
import multiple from '@/examples/select/demos/multiple.vue';
import noborder from '@/examples/select/demos/noborder.vue';
import popupProps from '@/examples/select/demos/popup-props.vue';
import prefix from '@/examples/select/demos/prefix.vue';
import remoteSearch from '@/examples/select/demos/remote-search.vue';
import size from '@/examples/select/demos/size.vue';
import status from '@/examples/select/demos/status.vue';

const mapper = {
  base,
  collapsed,
  creatable,
  customOptions,
  customSelected,
  filterable,
  disabled,
  group,
  labelInValue,
  max,
  multiple,
  noborder,
  popupProps,
  prefix,
  remoteSearch,
  size,
  status,
};

describe('Steps', () => {
  Object.keys(mapper).forEach((demoName) => {
    it(`Steps ${demoName} demo works fine`, () => {
      const wrapper = mount(mapper[demoName]);
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
