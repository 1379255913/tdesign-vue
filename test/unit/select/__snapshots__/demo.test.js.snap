// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select base demo works fine 1`] = `
<div>
  <div
    class="t-select-wrap"
    style="width: 200px;"
  >
    <div
      class="t-popup-reference t-select-popup-reference"
    >
      <transition-stub
        appear="true"
        name="t-popup_animation"
      >
        <div
          aria-hidden="true"
          class="t-popup"
          role="tooltip"
          style="display: none;"
        >
          <div
            class="t-popup-content  t-select-dropdown narrow-scrollbar"
          >
            <div>
              <ul
                class="t-select-create-option"
                style="display: none;"
              >
                <li
                  class="t-select-option t-size-m t-select-create-option-special"
                  style="display: none;"
                  title=""
                >
                  <span />
                </li>
              </ul>
              <li
                class="t-select-empty t-size-m"
              >
                暂无数据
              </li>
              <span
                style="display: none;"
              >
                <li
                  class="t-select-option t-is-selected t-size-m"
                  style="display: none;"
                  title="已选择的选项"
                >
                  <span>
                    
      已选择的选项
    
                  </span>
                </li>
                <li
                  class="t-select-option t-size-m"
                  style="display: none;"
                  title="短的选项二"
                >
                  <span>
                    
      短的选项二
    
                  </span>
                </li>
                <li
                  class="t-select-option t-size-m"
                  style="display: none;"
                  title="很长很长很很长很长很长的选项三"
                >
                  <span>
                    
      很长很长很很长很长很长的选项三
    
                  </span>
                </li>
              </span>
            </div>
          </div>
        </div>
      </transition-stub>
      <div
        class="t-select t-size-m"
      >
        <span
          class="t-select-selectedSingle"
          title="1"
        >
          1
        </span>
        <svg
          class="t-select-right-icon t-fake-arrow"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.75 5.7998L7.99274 10.0425L12.2361 5.79921"
            stroke="black"
            stroke-opacity="0.9"
            stroke-width="1.3"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
